README.md

```diff
@@ -1,9 +1,329 @@
+                expression.report(
+                    IncrementNotInConstantInt,
+                    "found `${+expression}` but @Increment expects a constant of type `Int`"
+                )
+            //language=kotlin
+            return "${+expression} + 1".call
+        }
+
+    }
+}
+```
+
+### Declaration generators
+Declaration generators enable the generation of top-level declarations and members during the compilation process. To use this feature, extend one of the `arrow.meta.Meta.Generate` sub interfaces, such as `arrow.meta.Meta..Generate.TopLevel` or `arrow.meta.Meta.Generate.Members`. These interfaces provide functions that allow you to generate various elements, including classes, functions, and properties.
+As an example, a declaration generator could generate a `product` function for a class annotated with the `@Product` annotation, which returns a list of pairs containing the names and values of the class's properties.
+
+```kotlin
+package arrow.meta.samples
+
+import arrow.meta.FirMetaContext
+import arrow.meta.Meta
+import org.jetbrains.kotlin.fir.declarations.*
+
+@Target(AnnotationTarget.CLASS)
+@Retention(AnnotationRetention.RUNTIME)
+@Meta
+annotation class Product {
+  companion object : Meta.Generate.Members.Functions {
+
+    override fun FirMetaContext.newFunctions(firClass: FirClass): Map<String, () -> String> =
+      mapOf(
+        "product" to {
+          //language=kotlin
+          """|
+             |fun product(): List<Pair<String, *>> = 
+             |  listOf(${propertiesOf(firClass) { """"${+it.name}" to this.${+it.name}""" }})
+          """.trimMargin()
+        }
+      )
+
+  }
+}
+```
+
+### Call Interceptors
+
+Call interceptors are a type of frontend transformer in Arrow Meta that allow you to intercept and modify function calls at compile time. They are defined by implementing the CallInterceptor interface and providing an intercept function that takes a list of arguments and a function to call with those arguments.
+
+To use a call interceptor, you will need to annotate the function that you want to intercept with the corresponding annotation. The intercept function will then be called every time the function is called, allowing you to modify the arguments and return value of the function.
+
+Here is an example of how you might use a call interceptor to increment all integer arguments passed to a function:
+
+```kotlin
+package arrow.meta.samples
+
+import arrow.meta.Meta
+import kotlin.reflect.KClass
+
+@Target(AnnotationTarget.FUNCTION)
+@Retention(AnnotationRetention.RUNTIME)
+@Meta
+annotation class IncrementArgs
+
+companion object : Meta.CallInterceptor {
+
+  override val annotation: KClass<*> = IncrementArgs::class
+
+  override fun <In, Out> intercept(args: List<In>, func: (List<In>) -> Out): Out {
+    val newArgs = args.map {
+      when (it) {
+        is Int -> (it + 1) as In //increments by one if it's an int
```

Comment:
```suggestion
        is Int -> (it + 1) as Int //increments by one if it's an int
```

Typo? Should be Int?

Reply:
it's correct because it is casting to the type parameter `In`